File: apps/web/src/entities/labeling/linkageEvaluator.test.ts
import { describe, expect, it } from 'vitest';
import type { LinkageAtomicCondition, LinkageConditionGroup, LinkageConditionOp } from '../schema/schemaTypes';
import { evaluateLinkageCondition } from './linkageEvaluator';

function atomic(field: string, op: LinkageConditionOp, value?: unknown): LinkageAtomicCondition {
  return value === undefined ? { field, op } : { field, op, value };
}

function allOf(...conditions: LinkageAtomicCondition[]): LinkageConditionGroup {
  return { allOf: conditions };
}

function anyOf(...conditions: LinkageAtomicCondition[]): LinkageConditionGroup {
  return { anyOf: conditions };
}

describe('frontend linkage evaluator', () => {
  it('evaluates scalar equality without cross-type coercion', () => {
    const values = new Map<string, unknown>([
      ['textNumber', '1'],
      ['number', 1],
      ['bool', true],
    ]);

    expect(evaluateLinkageCondition(atomic('textNumber', 'eq', '1'), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('number', 'eq', 1.0), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('textNumber', 'eq', 1), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('bool', 'neq', false), values)).toBe(true);
  });

  it('treats empty driver values as false for non-empty operators', () => {
    const values = { blank: '', emptyList: [] };

    expect(evaluateLinkageCondition(atomic('missing', 'eq', 'x'), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('missing', 'neq', 'x'), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('blank', 'in', ['x']), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('emptyList', 'notIn', ['x']), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('missing', 'gt', 1), values)).toBe(false);
  });

  it('reuses P3a empty semantics', () => {
    const values = {
      blank: '',
      spaces: '   ',
      emptyList: [],
      emptyMap: {},
      text: 'x',
    };

    expect(evaluateLinkageCondition(atomic('missing', 'empty'), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('blank', 'empty'), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('emptyList', 'empty'), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('emptyMap', 'empty'), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('spaces', 'empty'), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('text', 'notEmpty'), values)).toBe(true);
  });

  it('evaluates membership for scalars and arrays by intersection', () => {
    const values = {
      single: 'manual',
      multi: ['alpha', 'beta'],
    };

    expect(evaluateLinkageCondition(atomic('single', 'in', ['manual', 'other']), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('single', 'notIn', ['other']), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('multi', 'in', ['beta', 'gamma']), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('multi', 'notIn', ['gamma']), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('multi', 'notIn', ['alpha']), values)).toBe(false);
  });

  it('evaluates finite numeric comparisons and rejects non-numeric values', () => {
    const values = {
      count: 10,
      decimal: 1.5,
      text: '10',
    };

    expect(evaluateLinkageCondition(atomic('count', 'gt', 5), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('count', 'gte', 10.0), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('decimal', 'lt', 2), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('decimal', 'lte', 1.5), values)).toBe(true);
    expect(evaluateLinkageCondition(atomic('text', 'gt', 5), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('count', 'gt', '5'), values)).toBe(false);
    expect(evaluateLinkageCondition(atomic('count', 'lt', Number.POSITIVE_INFINITY), values)).toBe(false);
  });

  it('evaluates one-level allOf and anyOf groups', () => {
    const values = { type: 'other', score: 7 };

    expect(evaluateLinkageCondition(allOf(atomic('type', 'eq', 'other'), atomic('score', 'gte', 5)), values)).toBe(true);
    expect(evaluateLinkageCondition(allOf(atomic('type', 'eq', 'other'), atomic('score', 'lt', 5)), values)).toBe(false);
    expect(evaluateLinkageCondition(anyOf(atomic('type', 'eq', 'manual'), atomic('score', 'gte', 5)), values)).toBe(true);
  });

  it('returns false for null or malformed conditions', () => {
    expect(evaluateLinkageCondition(null, { field: 'value' })).toBe(false);
    expect(evaluateLinkageCondition({} as LinkageConditionGroup, { field: 'value' })).toBe(false);
    expect(evaluateLinkageCondition({ field: 'field' } as LinkageAtomicCondition, { field: 'value' })).toBe(false);
    expect(evaluateLinkageCondition(atomic('field', 'neq'), { field: 'value' })).toBe(false);
    expect(evaluateLinkageCondition(atomic('field', 'notIn', 'not-array'), { field: 'value' })).toBe(false);
    expect(evaluateLinkageCondition(atomic('object', 'neq', 'value'), { object: { nested: 'value' } })).toBe(false);
  });
});

File: services/api/src/test/java/com/labelhub/api/module/submission/validation/LinkageEvaluatorTest.java
package com.labelhub.api.module.submission.validation;

import com.labelhub.api.generated.model.LinkageAtomicCondition;
import com.labelhub.api.generated.model.LinkageCondition;
import com.labelhub.api.generated.model.LinkageConditionGroup;
import com.labelhub.api.generated.model.LinkageConditionOp;
import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class LinkageEvaluatorTest {

    private final LinkageEvaluator evaluator = new LinkageEvaluator();

    @Test
    void evaluates_scalar_equality_without_cross_type_coercion() {
        Map<String, Object> values = Map.of(
                "textNumber", "1",
                "number", 1,
                "bool", true);

        assertThat(evaluator.evaluate(atomic("textNumber", LinkageConditionOp.EQ, "1"), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("number", LinkageConditionOp.EQ, 1.0d), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("textNumber", LinkageConditionOp.EQ, 1), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("bool", LinkageConditionOp.NEQ, false), values)).isTrue();
    }

    @Test
    void treats_empty_driver_values_as_false_for_non_empty_ops() {
        Map<String, Object> values = Map.of(
                "blank", "",
                "emptyList", List.of());

        assertThat(evaluator.evaluate(atomic("missing", LinkageConditionOp.EQ, "x"), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("missing", LinkageConditionOp.NEQ, "x"), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("blank", LinkageConditionOp.IN, List.of("x")), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("emptyList", LinkageConditionOp.NOTIN, List.of("x")), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("missing", LinkageConditionOp.GT, 1), values)).isFalse();
    }

    @Test
    void reuses_p3a_empty_semantics() {
        Map<String, Object> values = Map.of(
                "blank", "",
                "spaces", "   ",
                "emptyList", List.of(),
                "emptyMap", Map.of(),
                "text", "x");

        assertThat(evaluator.evaluate(atomic("missing", LinkageConditionOp.EMPTY, null), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("blank", LinkageConditionOp.EMPTY, null), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("emptyList", LinkageConditionOp.EMPTY, null), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("emptyMap", LinkageConditionOp.EMPTY, null), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("spaces", LinkageConditionOp.EMPTY, null), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("text", LinkageConditionOp.NOTEMPTY, null), values)).isTrue();
    }

    @Test
    void evaluates_membership_for_scalars_and_collections() {
        Map<String, Object> values = Map.of(
                "single", "manual",
                "multi", List.of("alpha", "beta"));

        assertThat(evaluator.evaluate(atomic("single", LinkageConditionOp.IN, List.of("manual", "other")), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("single", LinkageConditionOp.NOTIN, List.of("other")), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("multi", LinkageConditionOp.IN, List.of("beta", "gamma")), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("multi", LinkageConditionOp.NOTIN, List.of("gamma")), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("multi", LinkageConditionOp.NOTIN, List.of("alpha")), values)).isFalse();
    }

    @Test
    void evaluates_numeric_comparisons_with_big_decimal_semantics() {
        Map<String, Object> values = Map.of(
                "count", 10,
                "decimal", new BigDecimal("1.50"),
                "text", "10");

        assertThat(evaluator.evaluate(atomic("count", LinkageConditionOp.GT, 5), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("count", LinkageConditionOp.GTE, 10.0d), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("decimal", LinkageConditionOp.LT, 2), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("decimal", LinkageConditionOp.LTE, new BigDecimal("1.5")), values)).isTrue();
        assertThat(evaluator.evaluate(atomic("text", LinkageConditionOp.GT, 5), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("text", LinkageConditionOp.LT, 5), values)).isFalse();
        assertThat(evaluator.evaluate(atomic("count", LinkageConditionOp.GT, "5"), values)).isFalse();
    }

    @Test
    void evaluates_one_level_all_of_and_any_of_groups() {
        Map<String, Object> values = Map.of("type", "other", "score", 7);

        assertThat(evaluator.evaluate(groupAllOf(
                atomic("type", LinkageConditionOp.EQ, "other"),
                atomic("score", LinkageConditionOp.GTE, 5)), values)).isTrue();
        assertThat(evaluator.evaluate(groupAllOf(
                atomic("type", LinkageConditionOp.EQ, "other"),
                atomic("score", LinkageConditionOp.LT, 5)), values)).isFalse();
        assertThat(evaluator.evaluate(groupAnyOf(
                atomic("type", LinkageConditionOp.EQ, "manual"),
                atomic("score", LinkageConditionOp.GTE, 5)), values)).isTrue();
    }

    @Test
    void returns_false_for_null_or_malformed_conditions() {
        assertThat(evaluator.evaluate(null, Map.of("field", "value"))).isFalse();
        assertThat(evaluator.evaluate(new UnknownCondition(), Map.of("field", "value"))).isFalse();

        LinkageAtomicCondition missingOp = new LinkageAtomicCondition();
        missingOp.setField("field");

        assertThat(evaluator.evaluate(missingOp, Map.of("field", "value"))).isFalse();
        assertThat(evaluator.evaluate(atomic("field", LinkageConditionOp.NEQ, null), Map.of("field", "value"))).isFalse();
        assertThat(evaluator.evaluate(atomic("field", LinkageConditionOp.NOTIN, "not-array"), Map.of("field", "value"))).isFalse();
        assertThat(evaluator.evaluate(atomic("object", LinkageConditionOp.NEQ, "value"), Map.of("object", Map.of("nested", "value")))).isFalse();
    }

    private static LinkageAtomicCondition atomic(String field, LinkageConditionOp op, Object value) {
        LinkageAtomicCondition condition = new LinkageAtomicCondition();
        condition.setField(field);
        condition.setOp(op);
        condition.setValue(value);
        return condition;
    }

    private static LinkageConditionGroup groupAllOf(LinkageAtomicCondition... conditions) {
        LinkageConditionGroup group = new LinkageConditionGroup();
        group.setAllOf(List.of(conditions));
        group.setAnyOf(null);
        return group;
    }

    private static LinkageConditionGroup groupAnyOf(LinkageAtomicCondition... conditions) {
        LinkageConditionGroup group = new LinkageConditionGroup();
        group.setAllOf(null);
        group.setAnyOf(List.of(conditions));
        return group;
    }

    private static final class UnknownCondition implements LinkageCondition {
    }
}
